### PR TITLE
Global placeholders

### PR DESCRIPTION
**Pull request by barnabwhy**
$:member, $:message, $:command, $:channel
$key (action specific)
$_key (any action, limited fields: text & variants, returns option; can be arbitrary)
$arg(n) (where n is the index of the arg, starting at 0)
HAS_ARGS condition, requires number of args
HAS_ITEM condition, requires item to have been purchased

### CODE REVIEW

**Comment by barnabwhy** [comment]
Beeg bug

**Comment by barnabwhy** [comment]
Fixing beeg bug now

**Comment by barnabwhy** [comment]
Bug fixed: Would permanently change the action with a modifier or placeholder. Copied object instead of referencing.

Missing feature: ARG_TEST emoji handling

File: src/bot.js [hunk 1]
```diff
@@ -100,6 +101,9 @@ client.on('message', async msg => {
                 channel: msg.channel,
                 member: member,
                 command: commandString,
+                args: args.filter(el => {
+                    return el != ''
+                }),
```
Comment (nanaian) [inline src/bot.js:106]: Can probably shorten this to
```suggestion
                args: args.filter(el => el != ''),
```

File: src/bot.js [hunk 2]
```diff
@@ -249,19 +257,20 @@ async function executeActionChain(actions, source) {
     }
 
     for (let idx = 0; idx < actions.length; idx++) {
-        const action = actions[idx]
+        var action = JSON.parse(JSON.stringify(actions[idx]))
```
Comment (nanaian) [inline src/bot.js:260]: Should be `const`

File: src/bot.js [hunk 3]
```diff
@@ -315,6 +324,7 @@ async function executeActionChain(actions, source) {
 
         await fn(source, makeResolvable(action), state).catch(err => {
             console.error(chalk.red(` error: ${err}`))
+            console.error(err.stack)
```
Comment (nanaian) [inline src/bot.js:327]: Make this red or remove it

File: src/bot.js [hunk 4]
```diff
@@ -361,6 +371,16 @@ function makeResolvable(map) {
...
 
     return {
+        getKeys() {
+            return Object.keys(map)
+        },
+
+        getString(key) {
+            const value = safeToString(resolveKey(key))
+
+            return value
```
Comment (nanaian) [inline src/bot.js:381]:
```suggestion
            return safeToString(resolveKey(key))
```

File: src/bot.js [hunk 5]
```diff
@@ -443,6 +464,72 @@ function makeResolvable(map) {
         },
     }
 }
+function safeToString(x) {
```
Comment (nanaian) [inline src/bot.js:467]: Should be an empty line of padding before this function

File: src/bot.js [hunk 6]
```diff
@@ -443,6 +464,72 @@ function makeResolvable(map) {
...
+            return '[Object object]'
+        case 'function':
+            return '[Function function]'
+        default:
+            return x + ''
+    }
+}
+
+const placeholdersInOpts = (opts, source) => {
+    var newOpts = opts
```
Comment (nanaian) [inline src/bot.js:479]: `const`

File: src/bot.js [hunk 7]
```diff
@@ -443,6 +464,72 @@ function makeResolvable(map) {
...
+
+const handlePlaceholders = (str, objs = {}) => {
+    if (objs.source.args) str = replaceArgPlaceholders(str, objs.source.args)
+    if (objs.source) str = replaceEventPlaceholders(str, objs.source)
+    if (objs.opts) str = replaceOptsPlaceholders(str, objs.opts)
+    return str
+}
+
+const replaceArgPlaceholders = (str, args) => {
+    for (var i = 0; i < args.length; i++) {
```
Comment (nanaian) [inline src/bot.js:506]:
```suggestion
    for (let i = 0; i < args.length; i++) {
```

File: src/bot.js [hunk 8]
```diff
@@ -443,6 +464,72 @@ function makeResolvable(map) {
...
+    var keys = Object.keys(opts)
+    keys.forEach(key => {
+        if (safeToString(opts[key]) == '[Object object]') return
+        const re = new RegExp(escapeRegexSpecialChars('$_' + key), 'g')
+        str = str.replace(re, safeToString(opts[key]))
+    })
+    return str
+}
+
+const escapeRegexSpecialChars = str => str.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1')
```
Comment (nanaian) [inline src/bot.js:532]: All of these functions should be moved to `utils.js` and required at the top of `bot.js` instead.
Some of them are already in `utils.js` - why duplicate code?

File: src/conditions.js [hunk 9]
```diff
@@ -47,4 +56,53 @@ module.exports = {
...
+        if (opts.getText('value') == 'Role') {
+            let role
+            if (target.startsWith('<@&')) {
+                const roleId = target.substr(3).slice(0, -1)
+                role = guild.roles.cache.find(r => r.id === roleId)
+            }
+            return role != undefined
+        }
+        if (opts.getText('value') == 'Emoji') {
+            // Alex please help with this one
```
Comment (nanaian) [inline src/conditions.js:103]: Will do

File: src/bot.js (hunk 3 above)
Comment (barnabwhy) [reply src/bot.js:327]: will remove

File: src/bot.js (hunk 6 above)
Comment (barnabwhy) [reply src/bot.js:479]: i tested with const but it didnt work

File: src/conditions.js [hunk 10]
```diff
@@ -47,4 +56,50 @@ module.exports = {
...
+        }
+        if (opts.getText('value') == 'Role') {
+            let role
+            if (target.startsWith('<@&')) {
+                const roleId = target.substr(3).slice(0, -1)
+                role = guild.roles.cache.find(r => r.id === roleId)
+            }
+            return role != undefined
+        }
+        // Type is not handled
```
Comment (nanaian) [inline src/conditions.js:102]: Would rather throw in this case

Comment (barnabwhy) [reply src/conditions.js:102]: ok

Comment (barnabwhy) [reply src/conditions.js:102]: is done
